url8d3c4d17f5_catseye_JaC64_tgz-pJ8-JaC64J8/src/com/dreamfabric/jac64/C64Screen.java

```diff
@@ -0,0 +1,2025 @@
+      // Bus not available if sp0 or sp1 is painting
+      if (sprites[1].dma) {
+        cpu.baLowUntil = lastLine + VICConstants.BA_SP1;
+      }
+      break;
+    case 57:
+      // Paint border, check sprite for display and read sprite 0 data.
+      for (int i = 0, n = 8; i < n; i++) {
+        Sprite sprite = sprites[i];
+        if (sprite.dma)
+          sprite.painting = true;
+      }
+
+      drawBackground();
+      drawSprites();
+      mpos += 8;
+
+      
+      if (rc == 7) {
+        vcBase = vc;
+        gfxVisible = false;
+        if (BAD_LINE_DEBUG) {
+          monitor.info("#### RC7 ==> vc = " + vc + " at " + vbeam +
+              " vicCycle = " + vicCycle);
+          if (vc == 1000) {
+            monitor.info("--------------- last line ----------------");
+          }
+        }
+      }
+
+      if (badLine || gfxVisible) {
+        rc = (rc + 1) & 7;
+        gfxVisible = true;
+      }
+
+      if (sprites[0].painting) {
+        sprites[0].readSpriteData();
+      }
+
+      if (sprites[2].dma) {
+        cpu.baLowUntil = lastLine + VICConstants.BA_SP2;
+      }
+
+      break;
+    case 58:
+      drawBackground();
+      drawSprites();
+      mpos += 8;
+
+      break;
+    case 59:
+      drawBackground();
+      drawSprites();
+      mpos += 8;
+
+      if (sprites[1].painting) {
+        sprites[1].readSpriteData();
+      }
+      break;
+    case 60:
+      drawSprites();
+      break;
+    case 61:
+      if (sprites[2].painting) {
+        sprites[2].readSpriteData();
+      }
+      if (sprites[3].dma) {
+        cpu.baLowUntil = lastLine + VICConstants.BA_SP3;
+      }
+      break;
+    case 62:
+      // Should this be made??? or should sprite 0 be repaintable 
+      // same line?
+      // Reset sprites so that they can be repainted again...
+      for (int i = 0; i < sprites.length; i++) {
+        sprites[i].reset();
+      }
+      lastLine += VICConstants.SCAN_RATE;
+      // Update screen
+      if (updating) {
+        if (vPos == 285) {
+          mis.newPixels();
+          canvas.repaint();
+          actualScanTime = (actualScanTime * 9 + (int)
+              ((audioDriver.getMicros() - lastScan))) / 10;
+          lastScan = audioDriver.getMicros();
+          updating = false;
+        }
+      }
+      notVisible = false;
+      break;
+    }
+  }
+
+  // Used to draw background where either border or background should be
+  // painted...
+  private void drawBackground() {
+    int bpos = mpos;
+    int currentBg = borderState > 0 ? borderColor : bgColor; 
+    for (int i = 0; i < 8; i++) {
+      mem[bpos++] = currentBg;
+    }
+  }
+
+  /**
+   * <code>drawGraphics</code> - draw the VIC graphics (text/bitmap)
+   * Note that sprites are not drawn here... (yet?)
+   *
+   *
+   * @param mpos an <code>int</code> value representing position to
+   * draw the graphics from (already fixed with hscroll)
+   */
+  private final void drawGraphics(int mpos) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 34. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.